pkgs/args/lib/src/json_schema.dart

```diff
@@ -0,0 +1,611 @@
+/// - Constraints on the number of properties ([minProperties],
+///   [maxProperties]).
+/// - Constraints on property names ([propertyNames]).
+///
+/// See https://json-schema.org/understanding-json-schema/reference/object.html
+/// for more details on object schemas.
+///
+/// Example:
+///
+/// To define a schema for a product object that requires `productId` and
+/// `productName`, has an optional `price` (non-negative number) and optional
+/// `tags` (list of unique strings), and optional `dimensions` (an object with
+/// required numeric length, width, and height):
+///
+/// ```dart
+/// final productSchema = ObjectSchema(
+///   title: 'Product',
+///   description: 'Schema for a product object',
+///   required: ['productId', 'productName'],
+///   properties: {
+///     'productId': Schema.string(
+///       description: 'Unique identifier for the product',
+///     ),
+///     'productName': Schema.string(description: 'Name of the product'),
+///     'price': Schema.num(
+///       description: 'Price of the product',
+///       minimum: 0,
+///     ),
+///     'tags': Schema.list(
+///       description: 'Optional list of tags for the product',
+///       items: Schema.string(),
+///       uniqueItems: true,
+///     ),
+///     'dimensions': ObjectSchema(
+///       description: 'Optional product dimensions',
+///       properties: {
+///         'length': Schema.num(),
+///         'width': Schema.num(),
+///         'height': Schema.num(),
+///       },
+///       required: ['length', 'width', 'height'],
+///     ),
+///   },
+///   additionalProperties: false, // No other properties allowed beyond those defined
+/// );
+/// ```
+///
+/// This schema can then be used with the `validate` method to check if a given
+/// JSON-like map conforms to the defined structure.
+///
+/// For example, valid data might look like:
+///
+/// ```json
+/// {
+///   "productId": "ABC12345",
+///   "productName": "Super Widget",
+///   "price": 19.99,
+///   "tags": ["electronics", "gadget"],
+///   "dimensions": {"length": 10, "width": 5, "height": 2.5}
+/// }
+/// ```
+///
+/// And invalid data (e.g., missing productName, or an extra undefined
+/// property):
+/// ```json
+/// {
+///   "productId": "XYZ67890",
+///   "price": 9.99
+/// }
+/// ```
+///
+/// ```json
+/// {
+///   "productId": "DEF4567",
+///   "productName": "Another Gadget",
+///   "color": "blue" // Invalid if additionalProperties is false
+/// }
+/// ```
+extension type ObjectSchema.fromMap(Map<String, Object?> _value)
+    implements Schema {
+  factory ObjectSchema({
+    String? title,
+    String? description,
+    Map<String, Schema>? properties,
+    Map<String, Schema>? patternProperties,
+    List<String>? required,
+
+    /// Must be one of bool, Schema, or Null
+    Object? additionalProperties,
+    bool? unevaluatedProperties,
+    StringSchema? propertyNames,
+    int? minProperties,
+    int? maxProperties,
+  }) =>
```

Comment:
(Could also be redirecting constructor `: this.fromMap(...);`. No big difference.)